Fix parsing of time specification with trailing spaces

dankamongmen opened the pull request:
There exists a unit test for a time spec of ' -1 M ', but it was failing on both my Debian and Arch machines. This is due to the change from 2019-08-07, "Improve time arg parsing". It looks like that's where support for the (undocumented) colon syntax was added, but it broke this test.

I have fixed up time parsing so that all tests pass. I haven't updated the usage summary, because I'm not entirely sure of the intended colon syntax's semantics. The usage summary could stand an update for 'T', I am thinking, and the colon syntax could use some unit tests. But this gets things going once more.

zuckschwerdt commented:
Thanks! How embarrassing. This change before `str = endptr;` should suffice to properly allow trailing spaces:
```
// allow trailing whitespace
        while (*endptr == ' ' || *endptr == '\t')
            ++endptr;
```
What problem did you encounter that needed shifting the `break`?

I tested this with some unit test. But I'd rather see you get credit for this change :)

dankamongmen commented:
Heh, i would prefer that, also :). 

Inspecting things more closely, I understand your intent with the colons better, and agree the break ought be moved back to its original place. I've restored it, and added `__attribute__((fallthrough))`. This both makes the intent clear (it looked like an oversight to me, and that probably biased my analysis), and allows `-Wimplicit-fallthrough` should you ever wish to turn it on.

Your other code is, aside from the comment, exactly what I have, right?

dankamongmen commented:
Updated to move the `break;` back. Good catch, thanks! If you want me to rebase these into a single commit or anything, just let me know. You're of course welcome to do a squash-merge if you'd like.

dankamongmen commented:
I rebased to pull down your additional atoi_time() tests, and we're passing them all. I also added a check+`exit()` to `style-check`, because if compiled with NDEBUG (the default), the assert() doesn't have an effect, and we hit undefined behavior (probably SIGSEGV). Feel free to remove anything you don't like.

zuckschwerdt commented:
Good catch.
Yes, we usually squash commit.
Not sure if the attribute is portable (we target MSVC too), GCC also offers a `// fall through` comment to do the same.

dankamongmen commented:
I'd replace it with a comment if you're worried about MSVC. It's unfortunately non-trivial to do a compatibility layer via macros: https://lwn.net/SubscriberLink/794944/5a7c59f5d4b98cd1/.

dankamongmen commented:
I've replaced the __attribute__((fallthrough)) with a comment.
